Sources/Testing/ExitTests/WaitFor.swift

```diff
@@ -65,40 +103,18 @@ func wait(for pid: pid_t) async throws -> ExitCondition {
     source.resume()
   }
   withExtendedLifetime(source) {}
-  return try _blockAndWait(for: pid)
-#else
-  // On Linux, spin up a background thread and waitpid() there.
-  return try await withCheckedThrowingContinuation { continuation in
-    // Create a structure to hold the state needed by the thread, and box it
-    // as a refcounted pointer that we can pass to libpthread.
-    struct Context {
-      var pid: pid_t
-      var continuation: CheckedContinuation<ExitCondition, any Error>
-    }
-    let context = Unmanaged.passRetained(
-      Context(pid: pid, continuation: continuation) as AnyObject
-    ).toOpaque()
-
-    // The body of the thread: unwrap and take ownership of the context we
-    // created above, then call waitpid() and report the result/error.
-    let body: @convention(c) (UnsafeMutableRawPointer?) -> UnsafeMutableRawPointer? = { contextp in
-      let context = Unmanaged<AnyObject>.fromOpaque(contextp!).takeRetainedValue() as! Context
-      let result = Result { try _blockAndWait(for: context.pid) }
-      context.continuation.resume(with: result)
-      return nil
-    }
+#elseif os(Linux)
+  // Ensure the waiter thread is running.
+  _createWaitThread()
 
-    // Create the thread. We immediately detach it upon success to allow the
-    // system to reclaim its resources when done.
-    var thread = pthread_t()
-    switch pthread_create(&thread, nil, body, context) {
-    case 0:
-      _ = pthread_detach(thread)
-    case let errorCode:
-      continuation.resume(throwing: CError(rawValue: errorCode))
+  await withCheckedContinuation { continuation in
+    let oldContinuation = _childProcessContinuations.withLock { childProcessContinuations in
+      childProcessContinuations.updateValue(continuation, forKey: pid)
```

Comment:
If we add a call to `waitpid(WNOHANG)` here before adding the continuation to the dictionary, then `waitid(P_ALL)` will be forced to wait to acquire the lock either _before_ the call to `waitpid(WNOHANG)` or _after_ the continuation is added.